### ISSUE TITLE
Group classes should be part of the Topology object

### ISSUE THREAD

**Issue by dotsdl** [MDAnalysis/mdanalysis#1205]
Under the new topology system, it is possible to add new `TopologyAttr` objects to an existing `Topology`, which could give new features to the atoms, residues, or segments in the topology. Although the interface for this functionality at the `Universe` level is incomplete (can be done with `Universe.add_TopologyAttr`, but not obvious how to use it or well-documented), the idea here is additional information can be added to a `Topology` as desired. When new attributes are added, the `Universe`'s `Group` classes are also updated to reflect the new additions.

Because all topology information resides in the `Topology` object, it is possible for a single `Topology` to be shared by multiple `Universe` objects. Although rarely needed, for large systems (>10M atoms) this could allow for large savings on memory when multiple trajectory readers are needed. 

However, if a `TopologyAttr` is added to a `Universe` whose `Topology` is shared with other `Universe`s, the addition of this `TopologyAttr` won't be propagated to the `Group` classes of the other `Universe`s. To avoid this issue, I propose that the `Group` classes be associated with the `Topology` object, not the `Universe`. All `Universe` objects with the same `Topology` will then "see" the same updates to their `Group` objects automatically.

Thoughts? @richardjgowers what do you think?

**Comment by mnmelo** [comment]
I like the idea. Then `Universe._classes` and `Universe._class_bases` would move to `Universe.topology`. I think it makes more organization sense too.

I don't see any immediate problems and say let's do it!

**Comment by richardjgowers** [comment]
Yeah without digging into code, my feeling is it might be a lot neater. The
generation of classes could even be done by the Topology.

It might be we might have to make u.top a first class attribute (similar to
traj). But thats probably good anyway to show the data structures clearly.

On Sun, 12 Feb 2017, 9:16 p.m. mnmelo, <notifications@github.com> wrote:

I like the idea. Then Universe._classes and Universe._class_bases would
move to Universe.topology. I think it makes more organization sense too.

I don't see any immediate problems and say let's do it!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/MDAnalysis/mdanalysis/issues/1205#issuecomment-279249731>,
or mute the thread

**Comment by richardjgowers** [comment]
I put together a PR of what it would roughly look like.  I like having `Topology` store and manage the `Group` objects.

To play devils advocate, the alternative to @dotsdl 's original problem is to have some sort of master/slave arrangement with the Universes, where any cloned Universe instances don't have full control on adding/removing Attributes, so are essentially read only Universes (wrt topology, writing coordinates is fine).

**Comment by dotsdl** [comment]
Does this need to happen before we release 0.16.0? Don't think it's absolutely necessary, but haven't considered all possible problems of not getting it in?

**Comment by richardjgowers** [comment]
Should be fine after as its internal API anyway, rather than user facing. I
imagine this will clear up once we delve into serialisation more

**Comment by orbeckst** [comment]
@richardjgowers @mnmelo what do you want to do with this issue? Take off the "question" tag and think about implementing or closing for the time being?
